tritium-proxy/src/main/java/com/palantir/tritium/proxy/Proxies.java

```diff
@@ -59,27 +60,27 @@ static Class<?>[] interfaces(Class<?> iface, Class<?> delegateClass) {
         checkIsInterface(iface);
         Objects.requireNonNull(delegateClass, "delegateClass");
 
-        Set<Class<?>> interfaces = new LinkedHashSet<>();
-        interfaces.add(iface);
         if (delegateClass.isInterface()) {
-            interfaces.add(delegateClass);
-        } else {
-            interfaces.addAll(Arrays.asList(delegateClass.getInterfaces()));
+            if (iface.equals(delegateClass)) {
+                return new Class<?>[] {iface};
+            }
+            return new Class<?>[] {iface, delegateClass};
         }
 
-        checkAreAllInterfaces(interfaces);
-        return interfaces.toArray(new Class<?>[0]);
+        Class<?>[] delegateInterfaces = delegateClass.getInterfaces();
+        int expectedSize = delegateInterfaces.length + 1;
+        Set<Class<?>> interfaces = new LinkedHashSet<>((int) ((float) expectedSize / 0.75F + 1.0F));
+        interfaces.add(iface);
+        for (Class<?> possibleInterface : delegateInterfaces) {
+            interfaces.add(checkIsInterface(possibleInterface));
+        }
+        return interfaces.toArray(EMPTY_CLASS_ARRAY);
     }
 
-    static void checkIsInterface(Class<?> iface) {
+    static Class<?> checkIsInterface(Class<?> iface) {
         if (!iface.isInterface()) {
             throw new IllegalArgumentException(iface + " is not an interface");
```

Comment:
Could use safe-logging exceptions here, but not much difference either (risk exceeding inline threshold if we're worried about that)

Reply:
I think I may convert these to logsafe in another PR to remove the suppressions and migrate over.